lib/boundkeypair/join_state.go

```diff
@@ -0,0 +1,225 @@
+		Claims: &jwt.Claims{
+			// We'll reuse the challengeNotBeforeOffset here; the value is sane
+			// enough.
+			NotBefore: jwt.NewNumericDate(params.Clock.Now().Add(challengeNotBeforeOffset)),
+			IssuedAt:  jwt.NewNumericDate(params.Clock.Now()),
+			Issuer:    params.ClusterName,
+			Audience:  jwt.Audience{params.ClusterName},
+			Subject:   params.Token.Spec.BotName,
+
+			// Note: These documents aren't meant to expire, so no expiration is
+			// included. We may opt to trust (or not) a given document during
+			// verification based on its `iat` in the future.
+		},
+		BotInstanceID:    status.BoundBotInstanceID,
+		RecoverySequence: status.RecoveryCount,
+		RecoveryLimit:    spec.Recovery.Limit,
+		RecoveryMode:     spec.Recovery.Mode,
+	}
+
+	// Derive the key ID for inclusion in the header.
+	kid, err := libjwt.KeyID(signer.Public())
+	if err != nil {
+		return "", trace.Wrap(err, "generating key ID")
+	}
+
+	signingKey, err := libjwt.SigningKeyFromPrivateKey(signer)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	opts := (&jose.SignerOptions{}).WithType("JWT").WithHeader("kid", kid)
+	joseSigner, err := jose.NewSigner(signingKey, opts)
+	if err != nil {
+		return "", trace.Wrap(err, "creating signer")
+	}
+
+	serialized, err := jwt.Signed(joseSigner).Claims(state).CompactSerialize()
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	return serialized, nil
+}
+
+func verifyJoinStateInner(key crypto.PublicKey, parsed *jwt.JSONWebToken, params *JoinStateParams) (*JoinState, error) {
+	if params.Token.Status == nil || params.Token.Status.BoundKeypair == nil {
+		return nil, trace.BadParameter("invalid token status")
+	}
+
+	var document JoinState
+	if err := parsed.Claims(key, &document); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Note: We don't verify expiry here, only `iat` and `nbf`. These documents
+	// are meant to remain "valid" indefinitely and we decide to trust them (or
+	// not) at verification time.
+	// There are no time-based recovery restrictions yet, but we may opt to add
+	// some in the future based on the verified `iat` value of this JWT.
+	const leeway time.Duration = time.Minute
+	if err := document.Claims.ValidateWithLeeway(jwt.Expected{
+		Issuer:   params.ClusterName,
+		Audience: jwt.Audience{params.ClusterName},
+		Subject:  params.Token.Spec.BotName,
+		Time:     params.Clock.Now(),
+	}, leeway); err != nil {
+		return nil, trace.Wrap(err, "validating join state claims")
+	}
+
+	// Ensure the non-informational claims in the join state match what we
+	// expect.
+	var errors []error
+	if document.RecoverySequence != params.Token.Status.BoundKeypair.RecoveryCount {
+		errors = append(errors, trace.AccessDenied("recovery counter mismatch"))
+	}
+	if document.BotInstanceID != params.Token.Status.BoundKeypair.BoundBotInstanceID {
+		errors = append(errors, trace.AccessDenied("bot instance mismatch"))
+	}
+
+	if len(errors) > 0 {
+		return nil, trace.NewAggregate(errors...)
+	}
```

Comment:
Tip: If you pass `NewAggregate` an empty slice or a slice containing only `nil` values, it'll return `nil` - so you don't actually need this check, although it's also fine if you want to keep it for clarity.